src/block/Furnace.php

```diff
@@ -90,4 +94,83 @@ public function onScheduledUpdate() : void{
 			$world->scheduleDelayedBlockUpdate($this->position, 1); //TODO: check this
 		}
 	}
+
+	public function doHopperPull(Hopper $hopperBlock) : bool{
+		$currentTile = $this->position->getWorld()->getTile($this->position);
+		if(!$currentTile instanceof TileFurnace){
+			return false;
+		}
+
+		$tileHopper = $this->position->getWorld()->getTile($hopperBlock->position);
+		if(!$tileHopper instanceof TileHopper){
+			return false;
+		}
+
+		$sourceInventory = $currentTile->getInventory();
+		$targetInventory = $tileHopper->getInventory();
+
+		$itemStack = $sourceInventory->getResult();
+
+		if($itemStack->isNull()){
+			return false;
+		}
+
+		$singleItem = $itemStack->pop();
+
+		$sourceInventory->removeItem($singleItem);
+		$targetInventory->addItem($singleItem);
+
+		return true;
+	}
+
+	public function doHopperPush(Hopper $hopperBlock) : bool{
+		$currentTile = $this->position->getWorld()->getTile($this->position);
+		if(!$currentTile instanceof TileFurnace){
+			return false;
+		}
+
+		$tileHopper = $this->position->getWorld()->getTile($hopperBlock->position);
+		if(!$tileHopper instanceof TileHopper){
+			return false;
+		}
+
+		$sourceInventory = $tileHopper->getInventory();
+		$targetInventory = $currentTile->getInventory();
+
+		$hopperFacing = $hopperBlock->getFacing();
+
+		for($i = 0; $i < $targetInventory->getSize(); $i++){
+			$itemStack = $sourceInventory->getItem($i);
+
+			if($itemStack->isNull()){
+				continue;
+			}
+
+			$singleItem = $itemStack->pop();
+
+			if($hopperFacing === Facing::DOWN && $targetInventory->canAddSmelting($singleItem)){
+				$this->transferItem($sourceInventory, $targetInventory, $singleItem, FurnaceInventory::SLOT_INPUT);
+				return true;
+			}elseif($hopperFacing !== Facing::DOWN && $hopperFacing !== Facing::UP && $targetInventory->canAddFuel($singleItem)){
```

Comment:
Can be reduced by using ``Axis`` ?

Reply:
Not sure I understand this but I did make a commit to reduce the code here

Reply:
Instead of checking both DOWN and UP facing, it can be better to check an Axis, X ou Y